pkg/controllers/deprovisioning/controller.go

```diff
@@ -297,26 +301,40 @@ func (c *Controller) waitForDeletion(ctx context.Context, nodeClaim *v1beta1.Nod
 	}
 }
 
-func (c *Controller) setNodesUnschedulable(ctx context.Context, isUnschedulable bool, candidates ...*Candidate) error {
+// requireDisruptionNoScheduleTaint will either add/remove the karpenter.sh/disruption taint from the candidates.
+// This is used to enforce no taints at the beginning of deprovisioning, and
+// to add/remove taints while executing a deprovisioning action.
+func (c *Controller) requireDisruptionNoScheduleTaint(ctx context.Context, addTaint bool, nodes ...*state.StateNode) error {
 	var multiErr error
-	for _, cn := range candidates {
+	for _, n := range nodes {
 		node := &v1.Node{}
-		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: cn.Node.Name}, node); err != nil {
+		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: n.Name()}, node); client.IgnoreNotFound(err) != nil {
```

Comment:
```suggestion
		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: n.Node.Name}, node); client.IgnoreNotFound(err) != nil {
```

You should probably just directly grab the node name here just to ensure that you are getting the right name here. `n.Name()` _should_ resolve to the node name, but it has the potential to not to.